src/ucp/core/ucp_ep.c

```diff
@@ -3194,12 +3198,20 @@ ucs_status_t ucp_ep_do_uct_ep_keepalive(ucp_ep_h ucp_ep, uct_ep_h uct_ep,
     ucp_wireup_msg_t wireup_msg;
 
     ucs_assert(!(ucp_ep->flags & UCP_EP_FLAG_FAILED));
-    ucs_assert((rsc_idx == UCP_NULL_RESOURCE) ||
-               (ucp_worker_iface(ucp_ep->worker, rsc_idx)->attr.cap.flags &
-                UCT_IFACE_FLAG_EP_CHECK));
 
     if (!ucp_ep_is_am_keepalive(ucp_ep, rsc_idx)) {
-        return uct_ep_check(uct_ep, flags, comp);
+        if ((rsc_idx == UCP_NULL_RESOURCE) ||
+            ucs_test_all_flags(
+                    ucp_worker_iface(ucp_ep->worker, rsc_idx)->attr.cap.flags,
+                                     UCT_IFACE_FLAG_EP_CHECK)) {
+            return uct_ep_check(uct_ep, flags, comp);
+        } else {
+            ucs_assert(!(ucp_ep->flags & UCP_EP_FLAG_REMOTE_ID) ||
+                       (ucp_worker_iface(ucp_ep->worker,
+                                         rsc_idx)->attr.cap.flags &
+                        UCT_IFACE_FLAG_EP_KEEPALIVE));
+            return UCS_OK;
+        }
```

Comment:
1. Some checks here seem like a duplication of ucp_ep_is_am_keepalive. So can unite/move them inside that funciton.
2. I would expect we would not have a case where need to skip the keepalive altogether like we have in lines 3209-3213. why does it happen? pls add code comments since it's not obvious